### DISCUSSION TITLE
[Codeql] How to update the database when I minor change the source code.

### DISCUSSION THREAD

**Discussion by pcy190** [github/securitylab#219] [Q&A] [answered]
I've generate the database from my source code. And If I modify something to the source code, How could I **regenerate** the database? I've tried the `codeql database update` but it seems doesn't help.  
I now has to delete the original database and create the database again, which I think is a little bit inefficient. I wonder if Codeql has the command line for me to update the database just from the generated database.   
(So I could minor change the source code and test the query again conveniently)  
Thanks.

**Comment by p0** [comment, marked as answer]
Short answer: There's currently no better way than what you're already doing -- deleting the database and re-creating it.

Longer explanation: For compiled languages, creating a CodeQL database observes a full build in order to interpret the source code correctly. It wouldn't be possible to update the information in the database correctly based purely on a source file diff -- for example, we wouldn't know what binaries that source file is linked to, what platforms/architectures it is compiled for, or what its compile-time dependencies are.

For interpreted languages (like JavaScript and Python), you could in principle use an "extractor cache" to avoid re-processing unchanged files, but the extraction phase tends to be much less of a bottleneck there.

What particular language are you working with here?

**Reply by pcy190** [reply]
Thank you for your reply. Besides, I'm working with C++ and have this question previously.

**Reply by p0** [reply]
Just to help us keep track of demand for this: What build system is your project using? Does it have a reliable incremental rebuild?

**Reply by pcy190** [reply]
Using cmake and clang. Seems the clang doesn't have incremental rebuild feature.  By the way, keep track of the unchanged output module and only recalculate the changed module like cmake, is a good way to incremental feature.

**Reply by pcy190** [reply]
And under Linux. As far as I know, visual studio cc compiler have incremental link feature, it has the LTO file which help to incremental rebuild.

**Comment by SongOfYouth** [comment]
Is it possible to make the database like git(see https://github.com/dolthub/dolt)? thus we can query it anyway we want. @p0

**Comment by dnovatchev** [comment]
A year and a half after this issue has been raised, is there any new development/progress?

**Reply by intrigus-lgtm** [reply]
There is a GitHub Next project for incremental CodeQL, but I don't know of any public alpha/beta/prototype for that:
https://githubnext.com/projects/incremental-codeql/
